### PR TITLE
expression,planner/core: support unix_timestamp() function in partition pruning

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve? 

```
CREATE TABLE quarterly_report_status (
    report_id INT NOT NULL,
    report_status VARCHAR(20) NOT NULL,
    report_updated TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP) 
PARTITION BY RANGE (UNIX_TIMESTAMP(report_updated)) (
    PARTITION p0 VALUES LESS THAN (UNIX_TIMESTAMP('2008-01-01 00:00:00')),
    PARTITION p1 VALUES LESS THAN (UNIX_TIMESTAMP('2008-04-01 00:00:00')),
    PARTITION p2 VALUES LESS THAN (UNIX_TIMESTAMP('2010-01-01 00:00:00')),
    PARTITION p3 VALUES LESS THAN (MAXVALUE)
);
```

Expected:

```
mysql> explain select * from quarterly_report_status where report_updated < '2007-05-01 00:00:00';
+---------------------+----------+------+------------------------------------------------------------------------------------------------+
| id                  | count    | task | operator info                                                                                  |
+---------------------+----------+------+------------------------------------------------------------------------------------------------+
| TableReader_8       | 3323.33  | root | data:Selection_7                                                                               |
| └─Selection_7       | 3323.33  | cop  | lt(test.quarterly_report_status.report_updated, 2007-05-01 00:00:00.000000)                    |
|   └─TableScan_6     | 10000.00 | cop  | table:quarterly_report_status, partition:p0, range:[-inf,+inf], keep order:false, stats:pseudo |
+---------------------+----------+------+------------------------------------------------------------------------------------------------+
3 rows in set (0.00 sec)
```

Get:
```
mysql> explain select * from quarterly_report_status where report_updated < '2007-05-01 00:00:00';
+--------------------------+----------+------+------------------------------------------------------------------------------------------------+
| id                       | count    | task | operator info                                                                                  |
+--------------------------+----------+------+------------------------------------------------------------------------------------------------+
| Union_10                 | 13293.33 | root |                                                                                                |
| ├─TableReader_13         | 3323.33  | root | data:Selection_12                                                                              |
| │ └─Selection_12         | 3323.33  | cop  | lt(test.quarterly_report_status.report_updated, 2007-05-01 00:00:00.000000)                    |
| │   └─TableScan_11       | 10000.00 | cop  | table:quarterly_report_status, partition:p0, range:[-inf,+inf], keep order:false, stats:pseudo |
| ├─TableReader_16         | 3323.33  | root | data:Selection_15                                                                              |
| │ └─Selection_15         | 3323.33  | cop  | lt(test.quarterly_report_status.report_updated, 2007-05-01 00:00:00.000000)                    |
| │   └─TableScan_14       | 10000.00 | cop  | table:quarterly_report_status, partition:p1, range:[-inf,+inf], keep order:false, stats:pseudo |
| ├─TableReader_19         | 3323.33  | root | data:Selection_18                                                                              |
| │ └─Selection_18         | 3323.33  | cop  | lt(test.quarterly_report_status.report_updated, 2007-05-01 00:00:00.000000)                    |
| │   └─TableScan_17       | 10000.00 | cop  | table:quarterly_report_status, partition:p2, range:[-inf,+inf], keep order:false, stats:pseudo |
| └─TableReader_22         | 3323.33  | root | data:Selection_21                                                                              |
|   └─Selection_21         | 3323.33  | cop  | lt(test.quarterly_report_status.report_updated, 2007-05-01 00:00:00.000000)                    |
|     └─TableScan_20       | 10000.00 | cop  | table:quarterly_report_status, partition:p3, range:[-inf,+inf], keep order:false, stats:pseudo |
+--------------------------+----------+------+------------------------------------------------------------------------------------------------+
13 rows in set (0.00 sec)
```

Partition pruning doesn't work using the `unix_timestamp()` function.

### What is changed and how it works?

Add `unix_timestamp` to monotoneIncFuncs list.
We can infer the rule `f(x) > xxx and x < yyy` if f is monotonously increasing.

### Check List 

Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)

### CODE REVIEW

**Comment by tiancaiamao** [comment]
PTAL @XuHuaiyu @eurekaka

**Comment by tiancaiamao** [comment]
Related issue https://github.com/pingcap/tidb/issues/12028

File: expression/constant_test.go [hunk 1]
```diff
@@ -56,14 +56,22 @@ func newLonglong(value int64) *Constant {
 }
 
 func newDate(year, month, day int) *Constant {
+	return newTimeConst(year, month, day, 0, 0, 0, mysql.TypeDate)
+}
+
+func newTimestamp(yy, mm, dd, hh, min, ss int) *Constant {
+	return newTimeConst(yy, mm, dd, hh, min, ss, mysql.TypeTimestamp)
```
Comment (Fatpa) [inline expression/constant_test.go]: The arguments `hh, min, ss` dosen't use in the `newTimeConst` function, why still parse them into the function??

Comment (tiancaiamao) [reply expression/constant_test.go]: A timestamp value has the `hh, min, ss` part, while a date value contains only the `yy, mm, dd` information. That's why the `newDate` function has fewer parameters than `newTimestamp`.

We pass 0 as arguments, but it doesn't means `hh, min, ss` would never be useful in the future.
@Fatpa

**Comment by tiancaiamao** [comment]
PTAL @XuHuaiyu @eurekaka @zz-jason

**Comment by cyliu0** [comment]
@tiancaiamao Do we need to cherry pick this one to release-3.1?

**Pull request merged**